### PR TITLE
Priming build and reload improvements.

### PR DESCRIPTION
**Pull request by sdedic**
There was a race condition with opening projects and subprojects with an empty `.m2/repository`. The issue was that when the main project was being primed, a file-based watcher recognized appearance of a dependency .jar and scheduled a reload of the project **before** the priming was completed.

Consider a multi-module Micronaut-based project:
```
[ ] example (parent = micronaut)
+--[ ] lib (parent = example)
+--[ ] app (parent = example)
```
Initially, when a subproject was opened, reading of the project failed as the immediate parent (root project) was locally present, but super-parent (micronaut-parent) was not resulting in properties not defined - the project read broke and was marked as a fallback, the same as the parent project. Good!

Because of the file-based trigger (micronaut-parent POM appears), the parent reloaded **while** other dependencies were still being downloaded. Some `jar` files already downloaded - and the `lib` project was able to resolve its dependencies to `jar` files. However Maven internally asks for `pom`-classified artifacts for the dependencies to compute (and download) a recursive closure. POMs were not available locally and were faked by `NbArtifactFixer`. However the information that Maven model resolver got incomplete dependency info was not recorded anywhere - those `Artifact` objects are used internally by Maven and do not appear in the `ProjectBuildingResult`.

So the basic fix has two parts:
- support **postponing** of project reload until after (some) project operation completion. In this case, sanity build (priming). File-based trigger will still trigger project reload, but that will happen only after the sanity build completes. This alone should ensure that child projects will (recursively) reload only after all artifacts are fetched.
- monitor **and report** nonlocal faked artifacts. Such projects will be marked as `incomplete`. After parent loads, all `incomplete` (not just fallbacks) projects will be reloaded, just in case, as parent's build should build all included modules.

References to (internal) `MavenProjectCache` replaced by `NbMavenProject` when possible.  Detailed logging added. Used `getPartialProject` to potentially unwrap at least some project structure after a failed load.

I've added basic unit tests to ensure that if a subproject primes, it becomes valid. And if parent primes, the subproject also becomes valid.

### CODE REVIEW

File: java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java
Comment (sdedic) [inline java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java:377]: Former call to `DefaultProjectBuilder.build` does not contain error handling / processing as implemented in `MavenProjectCache`. The other branch works with (nonlocal) artifacts, it is still TBD but serves mainly as a fallback.

File: java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java
Comment (sdedic) [inline java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java:580]: getMavenProject never returns `null`, but a fallback can be returned.

File: java/maven/src/org/netbeans/modules/maven/execute/ReactorChecker.java
Comment (sdedic) [inline java/maven/src/org/netbeans/modules/maven/execute/ReactorChecker.java:97]: This is important, as a subproject cannot be typically read with empty Maven repository. But the partial project contains a correct parent reference.

**Review by MartinBalin — APPROVED** [review]

**Comment by mbien** [comment]
just a general comment: I can't remember I ever primed anything using NetBeans (I am using the maven support since ea days). I press build and its ready (this works even if the build would fail, unless the pom is broken). I am not completely convinced that this is still a useful action, given that NB uses the local maven repo, project sources and target folder directly (I think some other IDEs have to prime since they have their own internal build representation, but this isn't the case here).

**Comment by sdedic** [comment]
> just a general comment: I can't remember I ever primed anything using NetBeans (I am using the maven support since ea days). [...] (I think some other IDEs have to prime since they have their own internal build representation, but this isn't the case here).

You probably didn't -- since the process happens automatically, if you open a Maven project with `Trust Project Build Script` enabled (default: on). Priming is also usually not necessary if the referenced artifacts are already present in local repository ... which is a typical situation on developer's machine that is used to develop a number of projects.

The second thing is that headless LSP usage of NB has some initialization paths not well reachable through NetBeans UI. Our colleagues also use 'throwaway' VMs to run the NBLS servers, so some of the scenarios come from using a pristine machine -- from the NB IDE point of view, they're testing for our first-time users/adopters.

**Review by dbalek — APPROVED** [review]
Thanks.

**Comment by neilcsmith-net** [comment]
> You probably didn't -- since the process happens automatically, if you open a Maven project with Trust Project Build Script enabled (default: on).

Sorry, where have we got automatic priming happening without user interaction?  That's potentially an issue.  As is treating Maven build scripts as trusted by default.

**Comment by mbien** [comment]
@sdedic thanks for explaining, some comments:
> You probably didn't -- since the process happens automatically,

I saw a few situations where NB asked me to prime something, I am pretty sure there is a notification bubble for that, and priming happens also at some point in the "resolve problems dialog", we had even an issue filed where a user used the "reload pom" action do download dependencies (!) (#6223), which I assume is running the priming code.

Why does NB not simply ask the user to build the project using the regular maven toolchain outside of the embedder? This could cause additional problem in future, e.g: #6190

I think mvn could also run `mvn dependency:go-offline` or something similar?


> BTW - in some of the later PRs I hope I will be able to avoid running mvn install

thats good since installing deps should not happen without user interaction IMO


(if the LSP server has special requirements - thats ok to me, esp headless mode is a completely different situation)

**Comment by neilcsmith-net** [comment]
> LSP case primes the projects as they are opened (similar to project open dialog). 

But what's the default in the VSCode case?  There shouldn't be automatic priming anywhere without the user opting in to that.

**Comment by dbalek** [comment]
> > LSP case primes the projects as they are opened (similar to project open dialog).
> 
> But what's the default in the VSCode case? There shouldn't be automatic priming anywhere without the user opting in to that.

In VSCode, whenever a new project is opened in workspace, user is asked to decide whether code in the project folder can be executed by VS Code and extensions without further explicit approval - see https://code.visualstudio.com/docs/editor/workspace-trust

**Comment by neilcsmith-net** [comment]
OK, thanks @dbalek - that's good to know they're covering this here for us.  Just concerned about ensuring we don't get automatic code execution anywhere with eg. `mvnw` in place.

**Comment by mbien** [comment]
just to show that I am not crazy, the bubble does exist :)
prime-bubble
(happened during manual testing of a different issue, I resolved it by building the project)

File: java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java
Comment (mbien) [inline java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java:615]: this looks a bit dangerous, the task is already running when passed into this method, which means that there is a point in time when the task is active without being marked as blocker.
```
((NbMavenProjectImpl)this.nbproject).scheduleProjectOperation(MavenModelProblemsProvider.RP.post(toRet));
```
Should this method have a RP param and actually schedule the tasks _inside_ the method? (+ throw exception when the task is already running).

Comment (sdedic) [reply java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java:615]: Good point. I'm glad that this is not a public API ... will follow the advice.

**Comment by sdedic** [comment]
Locally squashed.

File: java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java
Comment (mbien) [inline java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java:639]: Ok the events are fired even if the listener is added _after_ a task finishes - thats good. I probably would leave the delay parameter out, since this blocks the reload queue while doing nothing if I understand this correctly.

Are project ops allowed to run in parallel? If yes then `reloadBlocker` won't be sufficient as marker, since there can be multiple blockers at the same time and they can finish in arbitrary order.

I wonder if this would be simpler with a queue instead of events.

Comment (sdedic) [reply java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java:639]: > Are project ops allowed to run in parallel? If yes then `reloadBlocker` won't be sufficient as marker, since there can be multiple blockers at the same time and they can finish in arbitrary order.
> 

Who knows, I've just invented that concept ... but I'd say yes: there may be many operations having known (destructive :)) impact on loaded metadata, so it makes sense to delay the reloads.

Re. sufficiency: Pls. elaborate, I am not sure I've got the point.

> I wonder if this would be simpler with a queue instead of events. Since this listener chain essentially imitates a queue.

Pls. elaborate -- do you think something like:
- have a List<Task>, each having a TaskListener
- listener removes its Task from list, checks emptiness and if so, schedules a reload ?

Comment (mbien) [reply java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java:639]: Lets say one thread runs the ops sequentially, then tasks could be simply put into a single queue: `op0, op1, reload, op2`

This could even optimize the queue by only allowing one reload. If a second one is posted the other reload op is removed, so that it is always at the end of the queue.

If ops run in parallel however:
```
|--op0------------|
|--op1----|
|--op2-------|
```
and someone schedules a reload, which task is the blocker? This is difficult to track with an event system and one single variable.

Comment (mbien) [reply java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java:639]: i can take a closer look at it this weekend. but you could also queue all tasks and also the reload task on one sequential RP with 1 thread. It is probably the safest and simplest approach. If this turns out to be a performance bottleneck we can think about parallelizing this later.

Comment (mbien) [reply java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java:639]: added a commit on top of this PR: https://github.com/mbien/netbeans/commit/7d40cf67b0b2dd4077096d02bc4d067a4e57f5e7

While doing that i noticed another thing what is likely a bug: the reload task is set to `finished` during construction, but there are places in the code which check if its finished, for example `getFreshOriginalMavenProject` which would essentially always return true and the else branch would never run.

(I don't know all the details of NB's concurrency classes, so this might just be me not understanding something, however I changed that in the commit too since i couldn't find the spot which would set the task to not finished when it starts running)

**Comment by sdedic** [comment]
Completely forgot about this, apologies.
Two options: merge into the release "as is" & improve implementation in master as suggested by @mbien or plan to rebase this on delivery as soon as it is created. I don't think I'll make it in the remaining time with at least one re-approval.

The fixed bug is really ugly on throwaway VMs in virtual classes.

**Comment by mbien** [comment]
I am not sure if merging this so late is a good idea. We did something like this before and had to revert the PR since it caused side effects. There is always the next release.

also I would prefer to switch to a minimal reproducer if possible, instead of importing a full third party sample project. This is just my opinion, not a veto. Since this is about priming, i suppose the java code outside of placeholder classes isn't even needed.

**Comment by MartinBalin** [comment]
This PR is needed for 20 as it fixes misbehaviour of NetBeans in loading Maven project. The code in PR is tested. We will target this to `delivery` and then to NB20.

**Comment by neilcsmith-net** [comment]
Adding `do not merge` until ready, both branch and some issues.

I agree with @mbien about the concerns of merging late, and of the test project.  The binary Maven wrapper should not be in the repository IMO (switch to source-only wrapper) and I have concerns about the licensing of the test project files.  If this is truly needed, can it be published externally for consumption?

**Comment by sdedic** [comment]
> I agree with @mbien about the concerns of merging late, and of the test project. The binary Maven wrapper should not be in the repository (switch to source-only wrapper) and I have concerns about the licensing of the test project files. If this is truly needed, can it be published externally for consumption?

Re. the test project sources: these are just generated (including the LICENSE file) from `https://www.graal.cloud/gcn/launcher/`, which is freely available to everyone. Does it answer the licensing issue ?

**Comment by neilcsmith-net** [comment]
The test sources are marked (c) Oracle, so possibly need registering to be picked up in the source notice bundle (cc/ @matthiasblaesing - not sure how tests fit into that).

The Maven wrapper jar is a release blocker.

**Comment by sdedic** [comment]
> The test sources are marked (c) Oracle, so possibly need registering to be picked up in the source notice bundle (cc/ @matthiasblaesing - not sure how tests fit into that).
> 

OK, if that's an issue, should I create my own project from scratch just following the wizard's project structure ?

> The Maven wrapper jar is a release blocker.
Will try to change to source, or get rid of wrapper. Thanks.

**Comment by mbien** [comment]
> OK, if that's an issue, should I create my own project from scratch just following the wizard's project structure ?

for the purpose of testing the priming code, the most important part is the pom. I believe the source code can be mostly reduced to placeholder classes, the launchers are probably not needed at all most likely.

**Comment by sdedic** [comment]
I made an implementation overhaul of the PR.
1. implementation reworked: blocking tasks are kept in a queue. Queue's emptiness is checked upon completion of each task.
2. I've added quite a few tests to cover different scenarios of delayed / forced / no reload requests
3. I've recreated the test project data from scratch, with just a skeleton of project directories. None of the files should now contain copyright other than ASF or references to oracle except artifact names.

I will probably add back some basic classes in the future, and add more tests that will check diagnostics in the opened projects, i.e. no unresolved symbols in files.

**Review by MartinBalin — APPROVED** [review]

File: java/maven/test/unit/src/org/netbeans/modules/maven/NbMavenProjectImplTest2.java
Comment (mbien) [inline java/maven/test/unit/src/org/netbeans/modules/maven/NbMavenProjectImplTest2.java:1]: I don't think this test will run in CI since the default include is probably `*Test` or something similar

Comment (sdedic) [reply java/maven/test/unit/src/org/netbeans/modules/maven/NbMavenProjectImplTest2.java:1]: eeeh....

**Comment by sdedic** [comment]
ok, still some race condition, working on that

**Comment by sdedic** [comment]
temporarily removing labels so I do not run so many tests; enabling more logs. The race does not happen locally.

**Comment by sdedic** [comment]
@mbien -- I am surprised that Enterprise tests run in this label setup ... is it correct that labels are not checked for `enterprise-test` in `main.yml` ?

**Comment by mbien** [comment]
> @mbien -- I am surprised that Enterprise tests run in this label setup ... is it correct that labels are not checked for `enterprise-test` in `main.yml` ?

many tests are always on, esp those who don't use up many resources. E.g the PHP label toggles only the most expensive step in the php job.

If you want me to toggle more steps/jobs with labels - let me know. (the enterprise job used to be small initially and grew over time)

**Comment by sdedic** [comment]
OK... some explanation before I clean up the code again:
no race condition, but it turned out that the Micronaut4 tests can only run on JDK17+ - a Micronaut LifevcycleParticipant injected into Maven infrastructure in MN 4.0+ version is compiled for JDK17 and above. Then any Maven run on JDK11 fail during project model building phase, and even fails to download all artifacts -- which causes projects in NB to fail to load again, which ultimately asserts the test.

I would appreciate to run these tests (since MN4 has quite weird project structure with all the BOMs etc) - and in addition other Oracle work focuses on Micronaut ... and it's a reasonably complex application library to use as test data.

So I'd like to run build tools on **both** JDK11 and JDK17 as the behaviour of Gradle/Maven may differ on different JDKs.

**Comment by mbien** [comment]
> So I'd like to run build tools on both JDK11 and JDK17 as the behaviour of Gradle/Maven may differ on different JDKs.

this is fine but will increase the whack-a-mole factor since the job is one of the more unreliable ones. It already has retry wrappers for mx project etc.

once gradle supports JDK 21 this should be moved from 17 to 21, since if it works on 21 it will on 17 too with high probability

**Comment by sdedic** [comment]
Especially review commit 71b6dd4a26; I've realized that althgough I've defined `test.jms.flags` for java.mx.projects thathad been failing because of reflective access to un-opened package on JDK17, the flag was not passed to the unit test JVM. Seems that the place in `nbbuild/templates/common.xml`  is the rigt one - but please chedk.

**Review by dbalek — APPROVED** [review]

**Comment by mbien** [comment]
> Seems that the place in `nbbuild/templates/common.xml` is the rigt one - but please chedk.

I don't know either, but if it works it is probably the right one :), I am wondering though why it worked for the other modules, since this isn't the first one which sets the flags.

Should this be set too `<property name="test.jms.flags" value=""/>`? So that it doesn't fail if it is not set in the project.properties? Ant isn't very consistent when it can't resolve a property, sometimes it simply pastes the unresolved string as value.

**Comment by sdedic** [comment]
added a default property value.

**Comment by sdedic** [comment]
... and finally rebased on `deliver` for NB20

**Comment by neilcsmith-net** [comment]
The PR is still based on master though.  Please update the default branch if you want this in delivery.

> once gradle supports JDK 21 this should be moved from 17 to 21, since if it works on 21 it will on 17 too with high probability

Well, they'll only be one more release before 17 becomes our baseline!

**Comment by sdedic** [comment]
@neilcsmith-net done. All tests seem to pass. Please include in NB20.

**Comment by neilcsmith-net** [comment]
Thanks @sdedic Any concerns still from comments above @mbien ?  All being well, I will aim to merge tomorrow in time to sync for rc2.

Nitpick : there appears to be a left over `chmod` for `mvnw` in the build.xml.  I assume Ant doesn't care about this given everything built?!

**Comment by neilcsmith-net** [comment]
Don't mind!  Here would be better if you're planning that work on master any time in the next week though.

I did a quick search in the action output, but I guess it doesn't output that with the use of `-quiet`.

**Comment by neilcsmith-net** [comment]
OK, thanks @mbien Merging in time for 20-rc2.  Let's ensure that is well tested in this area.  It is always possible to revert for 20-rc3 in case of regressions.

**Comment by mbien** [comment]
project properties do not open #6622

**Pull request merged**
